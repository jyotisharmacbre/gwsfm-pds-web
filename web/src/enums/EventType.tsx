enum EventType {
  save,
  next,
  previous,
  none
}

export default EventType;
