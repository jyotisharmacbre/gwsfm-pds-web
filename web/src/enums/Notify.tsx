enum Notify {
  none,
  success,
  error
}

export default Notify;
