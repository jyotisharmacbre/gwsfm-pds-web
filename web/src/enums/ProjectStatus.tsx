enum ProjectStatus {
  InitialCustomerInquiry = 1,
  JA = 2,
  InReview = 3,
  BidLost = 6,
  OnHold = 7,
  OrderReceived = 5,
  JAApproved = 4,
  Live = 9,
  Completed = 10
}

export default ProjectStatus;
