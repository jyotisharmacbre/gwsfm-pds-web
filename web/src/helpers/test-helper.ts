import checkPropTypes from 'check-prop-types';

export const findByTestAtrr = (component: any, attr: string) => {
  const wrapper = component.find(`[data-test='${attr}']`);
  return wrapper;
};

export const checkProps = (component: any, expectedProps: any) => {
  const propsErr = checkPropTypes(
    component.propTypes,
    expectedProps,
    'props',
    component.name
  );
  return propsErr;
};
