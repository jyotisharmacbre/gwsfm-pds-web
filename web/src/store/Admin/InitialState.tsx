import { IAdminState } from './Types/IAdminState';
import { IAdminDefaults } from './Types/IAdminDefault';

export const initialState: IAdminState = {
	adminDefaultValues: [],
	error: ''
};
