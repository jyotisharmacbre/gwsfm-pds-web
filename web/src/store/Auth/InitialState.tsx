import { IAuth } from "./Types/IAuth";

export const initialState: IAuth= {
    token: null,
    name: null,
    firstName: null,
    email: null
};
