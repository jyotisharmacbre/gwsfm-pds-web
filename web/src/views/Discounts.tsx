import React, { Component, useEffect } from 'react';
import DiscountForm from '../components/Forms/Discount/DiscountForm';
import { FormattedMessage } from 'react-intl';
import { IDiscountActivity } from '../store/DiscountForm/Types/IDiscountActivity';
import EventType from '../enums/EventType';
import Notify from '../enums/Notify';
import * as actions from '../store/rootActions';
import { connect } from 'react-redux';
import { IState } from '../store/state';
import { toast } from 'react-toastify';
import { match } from 'react-router-dom';
import { History } from 'history';
import { ILookup } from '../store/Lookups/Types/ILookup';
import { ICurrency } from '../store/Lookups/Types/ICurrency';
import ProjectStatus from '../enums/ProjectStatus';
import { IDynamicContractCustomerData } from '../store/DynamicsData/Types/IDynamicData';
import { getClassNameForProjectStatus } from '../helpers/utility-helper';

interface IProps {
  match: match<{projectId:string}>;
  history : History;
}

interface IMapStateToProps {
  form: IDiscountActivity;
  notify: Notify;
  event: EventType;
  projectStatus: Array<ILookup>;
  currencies: Array<ICurrency> | null;
  currencyId: number;
  contractorId: string;
  otherCustomerName: string;
  status: number;
  dynamicsContractCustomerData: Array<IDynamicContractCustomerData>;
}

interface IMapDispatchToProps {
  getProjectStatus: () => void;
  discountFormAdd: (
    projectId: string,
    form: IDiscountActivity,
    event: EventType
  ) => void;
  discountFormEdit: (
    form: IDiscountActivity,
    event: EventType
  ) => void;
  resetDiscountState: () => void;
  getDiscountData: (projectId: string) => void;
  getAllCurrencies: () => void;
  getProjectDetail: (projectId: string) => void;
  getSubContractor: (projectId:string) => void;
    getPreliminaryDetails: (projectId: string) => void;
}

const Discounts: React.FC<IProps & IMapStateToProps & IMapDispatchToProps> = props => {
    const paramProjectId = props.match.params.projectId;
  
    useEffect(() => {
      window.scrollTo(0, 0);
      props.getProjectStatus();
      props.getAllCurrencies();
      if (paramProjectId != null && paramProjectId != '') {
        props.getProjectDetail(paramProjectId);
        props.getDiscountData(paramProjectId);
        props.getSubContractor(paramProjectId);
        props.getPreliminaryDetails(paramProjectId);
    }
  }, []);

  useEffect(() => {
    if (props.notify == Notify.success) {
      if (props.event == EventType.next) {
        toast.success('Data Saved Successfully');
        props.history.push(`/ReviewSubmit/${props.match.params.projectId}`);
      } else if (props.event == EventType.previous) {
        props.history.push(`/Subcontractor/${props.match.params.projectId}`);
      } else if (props.event == EventType.save) {
        toast.success('Data Saved Successfully');
      }
      props.resetDiscountState();
    }
  }, [props.notify, props.event]);

  const handlePrevious = () => {
    props.history.push(`/Subcontractor/${props.match.params.projectId}`);
  };

  const handleNext = (data: IDiscountActivity) => {
    data.discountId == ''
      ? props.discountFormAdd(paramProjectId, data, EventType.next)
      : props.discountFormEdit(data, EventType.next);
  };

  const handleSave = (data: IDiscountActivity) => {
    data.discountId == ''
      ? props.discountFormAdd(paramProjectId, data, EventType.save)
      : props.discountFormEdit(data, EventType.save);
  };

  return (
    <div className="container-fluid">
      <div
        data-test="dis_row_status"
        className={`${getClassNameForProjectStatus(props.status)} row`}>
        <div className="col-lg-12">
          <div className="custom-wrap discount_wrap">
            <div className="heading-subtitle">
              <h1>
                <FormattedMessage id="TITLE_JUSTIFICATION" />
              </h1>
              <p className="text-green">
                <FormattedMessage id="SUB_TITLE_DISCOUNTS" />
              </p>
            </div>
            <DiscountForm
              onNext={handleNext}
              onPrevious={handlePrevious}
              onSave={handleSave}
              projectstatus={props.projectStatus}
              currencies={props.currencies}
              currencyId={props.currencyId}
              contractorId={props.contractorId}
              otherCustomerName= {props.otherCustomerName} 
              projectId={props.match.params.projectId}
              dynamicsContractCustomerData={props.dynamicsContractCustomerData}/>
            </div>
          </div>
        </div>
      </div>
  );
};

const mapStateToProps = (state: IState) => ({
    form: state.discount.form,
    notify: state.discount.notify,
    event: state.discount.event,
    projectStatus: state.lookup.projectstatus,
    currencies: state.lookup.currencies,
    currencyId: state.project.form.currencyId,
    status:state.project.form.status,
    contractorId: state.project.form.contractorId,
    otherCustomerName: state.project.form.otherContractName,    
    dynamicsContractCustomerData: state.dynamicData.dynamicsContract
  })

  const mapDispatchToProps = dispatch => {
    return {
      getProjectStatus: () => dispatch(actions.getProjectStatus()),
      discountFormAdd: (projectId, form, event) =>
        dispatch(actions.discountFormAdd(projectId, form, event)),
      discountFormEdit: (form, event) =>
        dispatch(actions.discountFormEdit(form, event)),
      resetDiscountState: () =>
      dispatch(actions.resetDiscountState()),
      getAllCurrencies: () => dispatch(actions.getAllCurrencies()),
      getDiscountData: (projectId: string) =>
      dispatch(actions.getDiscountData(projectId)),
      getProjectDetail: projectId =>
      dispatch(actions.getProjectDetail(projectId)),
      getSubContractor: (projectId:string) =>
        dispatch(actions.getSubContractor(projectId)),
    getPreliminaryDetails: (projectId: string) =>
      dispatch(actions.getPreliminaryDetails(projectId)),
    };
}

  export default connect(mapStateToProps, mapDispatchToProps)(Discounts);
