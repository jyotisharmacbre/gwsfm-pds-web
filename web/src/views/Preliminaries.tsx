import React, { useEffect } from 'react';
import { connect } from 'react-redux';
import { IState } from '../store/state';
import { IPreliminariesComponentDetails } from '../store/Preliminaries/Types/IPreliminariesComponentDetails';
import * as actions from '../store/rootActions';
import PreliminaryForm from '../components/Forms/PreliminaryForm/PreliminaryForm';
import { IPreliminaries } from '../store/Preliminaries/Types/IPreliminaries';
import { convertIntoDatabaseModel } from "../store/Preliminaries/DataWrapper";
import { toast } from 'react-toastify';
import Notify from '../enums/Notify';
import { ILookup } from '../store/Lookups/Types/ILookup';
import { ICurrency } from '../store/Lookups/Types/ICurrency';
import { FormattedMessage } from 'react-intl';
import EventType from '../enums/EventType';
import CalculationsSummaryTable from '../components/Table/CalculationsSummaryTable';
import CalculationsSummaryType from '../enums/CalculationsSummaryType';
import { getPropertyName,getFilterElementFromArray } from '../helpers/utility-helper';
import { formValueSelector } from 'redux-form';
import Currency from '../store/Lookups/InitialState/Currency';
import { History } from 'history';
interface IMapStateToProps {
  preliminaryDetails: Array<IPreliminariesComponentDetails>;
  lookupData: Array<ILookup>;
  currencies: Array<ICurrency> | null;
  match: any;
  notify: Notify;
  event: EventType;
  currencyId: number;
  preliminaryForm:Array<IPreliminariesComponentDetails>;
  history:History;
}
interface IMapDispatchToProps {
  preliminaryAdd: (
    preliminaryDetails: Array<IPreliminaries>,
    event: EventType
  ) => void;
  preliminaryEdit: (
    preliminaryDetails: Array<IPreliminaries>,
    event: EventType
  ) => void;
  getPreliminaryDetails: (projectId: string) => void;
  updateInputField: (inputData: any) => void;
  getAllCurrencies: () => void;
  getProjectStatus: () => void;
  getProjectDetail: (projectId: string) => void;
}

const Preliminaries: React.FC<
  IMapStateToProps & IMapDispatchToProps
> = props => {
  let paramProjectId: string = '';
  const CurrencyObj = new Currency();
  let isLookupSessionExists: boolean = (sessionStorage.getItem("lookupData") != null && sessionStorage.getItem("lookupData") != undefined && sessionStorage.getItem("lookupData") != "")
  useEffect(() => {
    window.scrollTo(0, 0);
    paramProjectId=props.match.params.projectId;
    props.getProjectDetail(paramProjectId);
    props.getProjectStatus();
    props.getAllCurrencies();
    if (paramProjectId != null && paramProjectId != '' && paramProjectId != undefined && isLookupSessionExists) {
      props.getPreliminaryDetails(paramProjectId);
    }
  }, []);
  useEffect(() => {
    window.scrollTo(0, 0);
    if (props.notify == Notify.success) {
      toast.success('Data Saved Successfully');
      if (props.event == EventType.next) {
        props.history.push('/Subcontractor/' + props.match.params.projectId);
      }
    }
    else if (props.notify == Notify.error) {
      toast.error('Error occured.Please contact to administrator.');
    }

  }, [props.notify, props.event]);
  useEffect(() => {
    if (paramProjectId != null && paramProjectId != '' && paramProjectId != undefined && isLookupSessionExists) {
      props.getPreliminaryDetails(paramProjectId);
    }
  }, [props.lookupData]);
  const handleExpandAllEvent = () => {
    var element: any = document.getElementsByClassName('expandAll');
    for (let i = 0; i < element.length; i++) {
      element[i].classList.remove('hide');
      element[i].classList.add('show');
    }
  };
  const handleToggle = (id: string) => {
    var element: any = document.getElementById('collapse_' + id);
    if (element != null) {
      var isClassExists = element.classList.contains('show');
      if (isClassExists) {
        element.classList.add('hide');
        element.classList.remove('show');
      } else {
        element.classList.remove('hide');
        element.classList.add('show');
      }
    }
  };
  const handleSaveData = (
    saveAll: boolean,
    event: EventType,
    preliminaryDetails: any,
    index: number

  ) => {
    var editData: Array<IPreliminaries> = [];
    var saveData: Array<IPreliminaries> = [];
    paramProjectId = props.match.params.projectId;
    let preData: Array<IPreliminariesComponentDetails> = [];
    preData.push(preliminaryDetails.preliminaryDetails[index]);
    var preliminariesData: Array<IPreliminaries> = saveAll ?
      convertIntoDatabaseModel(preliminaryDetails.preliminaryDetails, paramProjectId) :
      convertIntoDatabaseModel(preData, paramProjectId)
    editData = preliminariesData.filter((data) => {
      return (data.TotalCost > 0 && data.PreliminaryId != '');
    })
    saveData = preliminariesData.filter((data) => {
      return (data.TotalCost > 0 && data.PreliminaryId == '');
    })
    if (editData.length > 0 && saveData.length > 0) { props.preliminaryEdit(editData, event); props.preliminaryAdd(saveData, event); }
    else if (editData.length > 0) { props.preliminaryEdit(editData, event); }
    else if (saveData.length > 0) { props.preliminaryAdd(saveData, event) }
    else { toast.error('No data changed to save.'); }
  };
  const handlePrevious = () => {
    props.history.push(`/Project/${props.match.params.projectId}`);
  };

  return (
    <div className="container-fluid">
      <div className=" row">
        <div className="col-lg-12">
          <form className="custom-wrap">
            <div className="heading-subtitle">
              <h1>
                <span className="d-md-block d-none">
                  <FormattedMessage id='TITLE_JUSTIFICATION'></FormattedMessage>
                </span>
                <span className="d-md-none"> <FormattedMessage id='TITLE_JUSTIFICATION_SHORT'></FormattedMessage></span>
              </h1>
              <p className="text-green"> <FormattedMessage id='PAGE_PRELIMINARY_TITLE'></FormattedMessage></p>
            </div>

            <div className="row">
              <div className="col-lg-12">
                <div className="row">
                  <div className="col-lg-8">
                    <div className="table-responsive">
                      <CalculationsSummaryTable
                      projectId={props.match.params.projectId}
                      name={CalculationsSummaryType.preliminary} 
                      preliminary={props.preliminaryForm}
                      currencySymbol={getFilterElementFromArray(
                        props.currencies,
                        getPropertyName(
                        CurrencyObj,
                        prop => prop.currencyId
                      ),
                        props.currencyId,
                        getPropertyName(
                        CurrencyObj,
                        prop => prop.currencySymbol
                      )
                      )}/>
                    </div>
                  </div>
                  <div className="col-lg-4">
                    <div className="text-right">
                      <button type="button" className="active fltRght" onClick={() => handleExpandAllEvent()}>EXPAND ALL</button>
                    </div>
                  </div>
                </div>
              </div>
            </div>

            <div>
              <PreliminaryForm
                onSave={handleSaveData}
                onPrevious={handlePrevious}
                onToggle={handleToggle}
                preliminariesDetails={props.preliminaryDetails}
                currencySymbol={getFilterElementFromArray(
                    props.currencies,
                    getPropertyName(
                    CurrencyObj,
                    prop => prop.currencyId
                  ),
                    props.currencyId,
                    getPropertyName(
                    CurrencyObj,
                    prop => prop.currencySymbol
                  )
                  )}
              />
            </div>
          </form>
        </div>
      </div>
    </div>
  );
};

const selector = formValueSelector('PreliminaryForm');


const mapStateToProps = (state: IState) => {
  return {
    preliminaryDetails: state.preliminary.preliminaryDetails,
    lookupData: state.lookup.projectstatus,
    currencies: state.lookup.currencies,
    notify: state.preliminary.notify,
    currencyId: state.project.form.currencyId,
    event: state.preliminary.event,
    preliminaryForm: selector(state, 'preliminaryDetails')
  };
};

const mapDispatchToProps = dispatch => {
  return {
    preliminaryAdd: (preliminaryDetails, event) =>
      dispatch(actions.preliminaryAdd(preliminaryDetails, event)),
    preliminaryEdit: (preliminaryDetails, event) =>
      dispatch(actions.preliminaryEdit(preliminaryDetails, event)),
    getProjectDetail: (projectId: string) =>
      dispatch(actions.getProjectDetail(projectId)),
    getPreliminaryDetails: (projectId: string) =>
      dispatch(actions.getPreliminaryDetails(projectId)),
    getAllCurrencies: () => dispatch(actions.getAllCurrencies()),
    getProjectStatus: () => dispatch(actions.getProjectStatus())
  };
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Preliminaries);
