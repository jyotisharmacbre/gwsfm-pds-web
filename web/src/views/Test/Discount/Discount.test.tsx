import React from 'react';
import { mount } from 'enzyme';
import Discounts from '../../Discounts';
import { findByTestAtrr } from '../../../helpers/test-helper';
import { IntlProvider } from 'react-intl';
import translations from '../../../Translations/translation';
import EventType from '../../../enums/EventType';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import {
	lookUpInitialState,
	customerEnquiryInitialState,
	dynamicsInitialState
} from '../Preliminaries/PreliminariesTestData';
import { discountInitialState } from './DiscountTestData';
import Notify from '../../../enums/Notify';
import { initialState as subcontractorInitialState } from '../../../store/SubContractor/InitialState';
import { initialState as preliminaryInitialState } from '../../../store/Preliminaries/InitialState';
import ProjectStatus from '../../../enums/ProjectStatus';

const mockStore = configureStore([]);
let store;
let wrapper;

const setUpStore = (lookUpInitialState, customerEnquiryInitialState) => {
	store = mockStore({
		discount: discountInitialState,
		lookup: lookUpInitialState,
		project: customerEnquiryInitialState,
		subContractor: subcontractorInitialState,
		preliminary: preliminaryInitialState,
		dynamicData: dynamicsInitialState,
		admin: { adminDefaultValues: [] }
	});
	store.dispatch = jest.fn();
};
const mountPreliminaryComponent = (Props) => {
	wrapper = mount(
		<Provider store={store}>
			<IntlProvider locale="en" messages={translations['en'].messages}>
				<Discounts {...Props} />
			</IntlProvider>
		</Provider>
	);
};
describe('Discount component test cases', () => {
	const Props: any = {
		form: discountInitialState.form,
		notify: Notify.none,
		projectId: '',
		event: EventType.none,
		projectStatus: [],
		currencies: [],
		currencyId: 1,
		customerName: 'test',
		otherCustomerName: 'test',
		status: 1,
		match: { params: { projectId: 1 } },
		history: { push: jest.fn() },
		getProjectStatus: jest.fn(),
		discountFormAdd: jest.fn(),
		discountFormEdit: jest.fn(),
		resetDiscountState: jest.fn(),
		getDiscountData: jest.fn(),
		getAllCurrencies: jest.fn()
	};
	beforeEach(() => {
		setUpStore(lookUpInitialState, customerEnquiryInitialState);
		mountPreliminaryComponent(Props);
	});

	it('defines the component', () => {
		expect(wrapper).toBeDefined();
	});
	it('should match the snapshot', () => {
		expect(wrapper).toMatchSnapshot();
	});
	it('should enable the form when state is not InReview', () => {
		expect(wrapper.find('.link_disabled').length).toEqual(0);
	});
	it('should not make preliminaries element into readonly if project status is not bidlost or onhold', () => {
		let container = findByTestAtrr(wrapper, 'dis_row_status').first();
		expect(container.hasClass('row')).toBe(true);
	});
	it('should make preliminaries element into readonly if project status is not bidlost or onhold', () => {
		customerEnquiryInitialState.form.status = ProjectStatus.BidLost;
		setUpStore(lookUpInitialState, customerEnquiryInitialState);
		mountPreliminaryComponent(Props);
		let container = findByTestAtrr(wrapper, 'dis_row_status').first();
		expect(container.hasClass('link_disabled')).toBe(true);
	});
	it('should disable the form when state is InReview', () => {
		let data = { ...customerEnquiryInitialState };
		data.form.status = ProjectStatus.InReview;
		setUpStore(lookUpInitialState, data);
		expect(wrapper.find('.link_disabled').length).toBeGreaterThan(0);
	});
	it('should make discount element into readonly if project status is order received', () => {
		customerEnquiryInitialState.form.status = ProjectStatus.OrderReceived;
		setUpStore(lookUpInitialState, customerEnquiryInitialState);
		mountPreliminaryComponent(Props);
		let container = findByTestAtrr(wrapper, 'dis_row_status').first();
		expect(container.hasClass('link_disabled')).toBe(true);
	});
});
