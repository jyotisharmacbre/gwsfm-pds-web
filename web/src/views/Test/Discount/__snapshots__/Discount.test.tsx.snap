// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Discount component test cases should match the snapshot 1`] = `ReactWrapper {}`;
